// vite-plugin-chrome-extension/lib/index-esm.js
import 'array-flat-polyfill';
import fs, { readFile, outputJson, readJSONSync } from 'fs-extra';
import flatten from 'lodash.flatten';
import path, { relative, join, basename } from 'path';
import cheerio from 'cheerio';
import prettier from 'prettier';
import { cosmiconfigSync } from 'cosmiconfig';
import { JSONPath } from 'jsonpath-plus';
import memoize from 'mem';
import slash from 'slash';
import glob from 'glob';
import get from 'lodash.get';
import diff from 'lodash.difference';
import Ajv from 'ajv';
import { rollup } from 'rollup';

const not = (fn) => (x) => !fn(x);

function isChunk(
  x,
) {
  return x && x.type === "chunk";
}

function isAsset(
  x,
) {
  return x.type === "asset";
}

function isString(x) {
  return typeof x === "string";
}

function isJsonFilePath(x) {
  return isString(x) && x.endsWith("json");
}

const normalizeFilename = (p) =>
  p.replace(/\.[tj]sx?$/, ".js");

/**
 * Update the manifest source in the output bundle
 */
const updateManifest = (
  updater

,
  bundle,
  handleError,
) => {
  try {
    const manifestKey = "manifest.json";
    const manifestAsset = bundle[manifestKey] ;

    if (!manifestAsset) {
      throw new Error(
        "No manifest.json in the rollup output bundle.",
      );
    }

    const manifest = JSON.parse(
      manifestAsset.source ,
    ) ;

    const result = updater(manifest);

    manifestAsset.source = JSON.stringify(result, undefined, 2);
  } catch (error) {
    if (handleError) {
      handleError(error.message);
    } else {
      throw error;
    }
  }

  return bundle;
};

function reduceToRecord(srcDir) {
  if (srcDir === null || typeof srcDir === "undefined") {
    // This would be a config error, so should throw
    throw new TypeError("srcDir is null or undefined");
  }

  return (
    inputRecord,
    filename,
  ) => {
    const name = relative(srcDir, filename)
      .split(".")
      .slice(0, -1)
      .join(".");

    if (name in inputRecord) {
      throw new Error(
        `Script files with different extensions should not share names:\n\n"${filename}"\nwill overwrite\n"${inputRecord[name]}"`,
      );
    }

    return { ...inputRecord, [name]: filename };
  };
}

const formatHtml = ($) =>
  prettier.format($.html(), { parser: "html" });

const loadHtml = (rootPath) => (
  filePath,
) => {
  const htmlCode = fs.readFileSync(filePath, "utf8");
  const $ = cheerio.load(htmlCode);

  return Object.assign($, { filePath, rootPath });
};

const getRelativePath = ({
  filePath,
  rootPath,
}) => (p) => {
  const htmlFileDir = path.dirname(filePath);

  let relDir;
  if (p.startsWith("/")) {
    relDir = path.relative(process.cwd(), rootPath);
  } else {
    relDir = path.relative(process.cwd(), htmlFileDir);
  }

  return path.join(relDir, p);
};

/* -------------------- SCRIPTS -------------------- */

const getScriptElems = ($) =>
  $("script")
    .not("[data-rollup-asset]")
    .not('[src^="http:"]')
    .not('[src^="https:"]')
    .not('[src^="data:"]')
    .not('[src^="/"]');

// Mutative action
const mutateScriptElems = ({
  browserPolyfill,
}) => (
  $,
) => {
  getScriptElems($)
    .attr("type", "module")
    .attr("src", (i, value) => {
      // FIXME: @types/cheerio is wrong for AttrFunction: index.d.ts, line 16
      // declare type AttrFunction = (i: number, currentValue: string) => any;
      // eslint-disable-next-line
      // @ts-ignore
      const replaced = value.replace(/\.[jt]sx?/g, ".js");

      return replaced;
    });

  if (browserPolyfill) {
    const head = $("head");
    if (
      browserPolyfill === true ||
      (typeof browserPolyfill === "object" &&
        browserPolyfill.executeScript)
    ) {
      head.prepend(
        '<script src="/assets/browser-polyfill-executeScript.js"></script>',
      );
    }

    head.prepend(
      '<script src="/assets/browser-polyfill.js"></script>',
    );
  }

  return $;
};

const getScripts = ($) =>
  getScriptElems($).toArray();

const getScriptSrc = ($) =>
  getScripts($)
    .map((elem) => $(elem).attr("src"))
    .filter(isString)
    .map(getRelativePath($));

/* ----------------- ASSET SCRIPTS ----------------- */

const getAssets = ($) =>
  $("script")
    .filter('[data-rollup-asset="true"]')
    .not('[src^="http:"]')
    .not('[src^="https:"]')
    .not('[src^="data:"]')
    .not('[src^="/"]')
    .toArray();

const getJsAssets = ($) =>
  getAssets($)
    .map((elem) => $(elem).attr("src"))
    .filter(isString)
    .map(getRelativePath($));

/* -------------------- css ------------------- */

const getCss = ($) =>
  $("link")
    .filter('[rel="stylesheet"]')
    .not('[href^="http:"]')
    .not('[href^="https:"]')
    .not('[href^="data:"]')
    .not('[href^="/"]')
    .toArray();

const getCssHrefs = ($) =>
  getCss($)
    .map((elem) => $(elem).attr("href"))
    .filter(isString)
    .map(getRelativePath($));

/* -------------------- img ------------------- */

const getImgs = ($) =>
  $("img")
    .not('[src^="http://"]')
    .not('[src^="https://"]')
    .not('[src^="data:"]')
    .toArray();

const getFavicons = ($) =>
  $('link[rel="icon"]')
    .not('[href^="http:"]')
    .not('[href^="https:"]')
    .not('[href^="data:"]')
    .toArray();

const getImgSrcs = ($) => {
  return [
    ...getImgs($).map((elem) => $(elem).attr("src")),
    ...getFavicons($).map((elem) => $(elem).attr("href")),
  ]
    .filter(isString)
    .map(getRelativePath($));
};

const isHtml = (path) => /\.html?$/.test(path);

const name$1 = "html-inputs";

/* ============================================ */
/*                  HTML-INPUTS                 */
/* ============================================ */

function htmlInputs(
    htmlInputsOptions,
    /** Used for testing */
    cache = {
        scripts: [],
        html: [],
        html$: [],
        js: [],
        css: [],
        img: [],
        input: [],
    } ,
) {
    return {
        name: name$1,
        cache,

        /* ============================================ */
        /*                 OPTIONS HOOK                 */
        /* ============================================ */

        options(options) {
            // srcDir may be initialized by another plugin
            const { srcDir } = htmlInputsOptions;

            if (srcDir) {
                cache.srcDir = srcDir;
            } else {
                throw new TypeError("options.srcDir not initialized");
            }

            // Skip if cache.input exists
            // cache is dumped in watchChange hook

            // Parse options.input to array
            let input;
            if (typeof options.input === "string") {
                input = [options.input];
            } else if (Array.isArray(options.input)) {
                input = [...options.input];
            } else if (typeof options.input === "object") {
                input = Object.values(options.input);
            } else {
                throw new TypeError(
                    `options.input cannot be ${typeof options.input}`,
                );
            }

            /* ------------------------------------------------- */
            /*                 HANDLE HTML FILES                 */
            /* ------------------------------------------------- */

            // Filter htm and html files
            cache.html = input.filter(isHtml);

            // If no html files, do nothing
            if (cache.html.length === 0) return options;

            // If the cache has been dumped, reload from files
            if (cache.html$.length === 0) {
                // This is all done once
                cache.html$ = cache.html.map(loadHtml(srcDir));

                cache.js = flatten(cache.html$.map(getScriptSrc));
                cache.css = flatten(cache.html$.map(getCssHrefs));
                cache.img = flatten(cache.html$.map(getImgSrcs));
                cache.scripts = flatten(cache.html$.map(getJsAssets));

                // Cache jsEntries with existing options.input
                cache.input = input.filter(not(isHtml)).concat(cache.js);

                // Prepare cache.html$ for asset emission
                cache.html$.forEach(mutateScriptElems(htmlInputsOptions));

                if (cache.input.length === 0) {
                    throw new Error(
                        "At least one HTML file must have at least one script.",
                    );
                }
            }

            // TODO: simply remove HTML files from options.input
            // - Parse HTML and emit chunks and assets in buildStart
            return {
                ...options,
                input: cache.input.reduce(
                    reduceToRecord(htmlInputsOptions.srcDir),
                    {},
                ),
            };
        },

        /* ============================================ */
        /*              HANDLE FILE CHANGES             */
        /* ============================================ */

        async buildStart() {
            const { srcDir } = htmlInputsOptions;

            if (srcDir) {
                cache.srcDir = srcDir;
            } else {
                throw new TypeError("options.srcDir not initialized");
            }

            const assets = [
                ...cache.css,
                ...cache.img,
                ...cache.scripts,
            ];

            assets.concat(cache.html).forEach((asset) => {
                this.addWatchFile(asset);
            });

            const emitting = assets.map(async (asset) => {
                // Read these files as Buffers
                const source = await readFile(asset);
                const fileName = relative(srcDir, asset);

                this.emitFile({
                    type: "asset",
                    source, // Buffer
                    fileName,
                });
            });

            cache.html$.map(($) => {
                const source = formatHtml($);
                const fileName = relative(srcDir, $.filePath);

                this.emitFile({
                    type: "asset",
                    source, // String
                    fileName,
                });
            });

            await Promise.all(emitting);
        },

        watchChange(id) {
            if (id.endsWith(".html") || id.endsWith("manifest.json")) {
                // Dump cache if html file or manifest changes
                cache.html$ = [];
            }
        },
    };
}

const code$5 = "(function () {\n\t'use strict';\n\n\tconst importPath = /*@__PURE__*/JSON.parse(\"%PATH%\") ;\n\n\timport(importPath);\n\n}());\n";

const cloneObject = (obj) => JSON.parse(JSON.stringify(obj));

const code$4 = "(function () {\n  'use strict';\n\n  function delay(ms) {\r\n    return new Promise((resolve) => {\r\n      setTimeout(resolve, ms);\r\n    });\r\n  }\n\n  function captureEvents(events) {\r\n    const captured = events.map(captureEvent);\r\n\r\n    return () => captured.forEach((t) => t());\r\n\r\n    function captureEvent(event) {\r\n      let isCapturePhase = true;\r\n\r\n      const callbacks = new Map();\r\n      const eventArgs = new Set();\r\n\r\n      // This is the only listener for the native event\r\n      event.addListener(handleEvent);\r\n\r\n      function handleEvent(...args) {\r\n        if (isCapturePhase) {\r\n          // This is before dynamic import completes\r\n          eventArgs.add(args);\r\n\r\n          if (typeof args[2] === \"function\") {\r\n            // During capture phase all messages are async\r\n            return true;\r\n          } else {\r\n            // Sync messages or some other event\r\n            return false;\r\n          }\r\n        } else {\r\n          // The callbacks determine the listener return value\r\n          return callListeners(...args);\r\n        }\r\n      }\r\n\r\n      // Called when dynamic import is complete\r\n      //  and when subsequent events fire\r\n      function callListeners(...args) {\r\n        let isAsyncCallback = false;\r\n        callbacks.forEach((options, cb) => {\r\n          // A callback error should not affect the other callbacks\r\n          try {\r\n            isAsyncCallback = cb(...args) || isAsyncCallback;\r\n          } catch (error) {\r\n            console.error(error);\r\n          }\r\n        });\r\n\r\n        if (!isAsyncCallback && typeof args[2] === \"function\") {\r\n          // We made this an async message callback during capture phase\r\n          //   when the function handleEvent returned true\r\n          //   so we are responsible to call sendResponse\r\n          // If the callbacks are sync message callbacks\r\n          //   the sendMessage callback on the other side\r\n          //   resolves with no arguments (this is the same behavior)\r\n          args[2]();\r\n        }\r\n\r\n        // Support events after import is complete\r\n        return isAsyncCallback;\r\n      }\r\n\r\n      // This function will trigger this Event with our stored args\r\n      function triggerEvents() {\r\n        // Fire each event for this Event\r\n        eventArgs.forEach((args) => {\r\n          callListeners(...args);\r\n        });\r\n\r\n        // Dynamic import is complete\r\n        isCapturePhase = false;\r\n        // Don't need these anymore\r\n        eventArgs.clear();\r\n      }\r\n\r\n      // All future listeners are handled by our code\r\n      event.addListener = function addListener(cb, ...options) {\r\n        callbacks.set(cb, options);\r\n      };\r\n\r\n      event.hasListeners = function hasListeners() {\r\n        return callbacks.size > 0;\r\n      };\r\n\r\n      event.hasListener = function hasListener(cb) {\r\n        return callbacks.has(cb);\r\n      };\r\n\r\n      event.removeListener = function removeListener(cb) {\r\n        callbacks.delete(cb);\r\n      };\r\n\r\n      event.__isCapturedEvent = true;\r\n\r\n      return triggerEvents;\r\n    }\r\n  }\n\n  function resolvePath(object, path, defaultValue) {\r\n    return path.split(\".\").reduce((o, p) => (o ? o[p] : defaultValue), object) ;\r\n  }\n\n  const eventPaths = /*@__PURE__*/JSON.parse(\"%EVENTS%\") ;\r\n  const importPath = /*@__PURE__*/JSON.parse(\"%PATH%\") ;\r\n  const delayLength = /*@__PURE__*/JSON.parse(\"%DELAY%\") ;\n\n  const events = eventPaths.map((eventPath) => resolvePath(chrome, eventPath));\r\n  const triggerEvents = captureEvents(events);\r\n\r\n  import(importPath).then(async () => {\r\n    if (delayLength) await delay(delayLength);\r\n\r\n    triggerEvents();\r\n  });\n\n}());\n";

const code$3 = "(function () {\n  'use strict';\n\n  function captureEvents(events) {\r\n    const captured = events.map(captureEvent);\r\n\r\n    return () => captured.forEach((t) => t());\r\n\r\n    function captureEvent(event) {\r\n      let isCapturePhase = true;\r\n\r\n      const callbacks = new Map();\r\n      const eventArgs = new Set();\r\n\r\n      // This is the only listener for the native event\r\n      event.addListener(handleEvent);\r\n\r\n      function handleEvent(...args) {\r\n        if (isCapturePhase) {\r\n          // This is before dynamic import completes\r\n          eventArgs.add(args);\r\n\r\n          if (typeof args[2] === \"function\") {\r\n            // During capture phase all messages are async\r\n            return true;\r\n          } else {\r\n            // Sync messages or some other event\r\n            return false;\r\n          }\r\n        } else {\r\n          // The callbacks determine the listener return value\r\n          return callListeners(...args);\r\n        }\r\n      }\r\n\r\n      // Called when dynamic import is complete\r\n      //  and when subsequent events fire\r\n      function callListeners(...args) {\r\n        let isAsyncCallback = false;\r\n        callbacks.forEach((options, cb) => {\r\n          // A callback error should not affect the other callbacks\r\n          try {\r\n            isAsyncCallback = cb(...args) || isAsyncCallback;\r\n          } catch (error) {\r\n            console.error(error);\r\n          }\r\n        });\r\n\r\n        if (!isAsyncCallback && typeof args[2] === \"function\") {\r\n          // We made this an async message callback during capture phase\r\n          //   when the function handleEvent returned true\r\n          //   so we are responsible to call sendResponse\r\n          // If the callbacks are sync message callbacks\r\n          //   the sendMessage callback on the other side\r\n          //   resolves with no arguments (this is the same behavior)\r\n          args[2]();\r\n        }\r\n\r\n        // Support events after import is complete\r\n        return isAsyncCallback;\r\n      }\r\n\r\n      // This function will trigger this Event with our stored args\r\n      function triggerEvents() {\r\n        // Fire each event for this Event\r\n        eventArgs.forEach((args) => {\r\n          callListeners(...args);\r\n        });\r\n\r\n        // Dynamic import is complete\r\n        isCapturePhase = false;\r\n        // Don't need these anymore\r\n        eventArgs.clear();\r\n      }\r\n\r\n      // All future listeners are handled by our code\r\n      event.addListener = function addListener(cb, ...options) {\r\n        callbacks.set(cb, options);\r\n      };\r\n\r\n      event.hasListeners = function hasListeners() {\r\n        return callbacks.size > 0;\r\n      };\r\n\r\n      event.hasListener = function hasListener(cb) {\r\n        return callbacks.has(cb);\r\n      };\r\n\r\n      event.removeListener = function removeListener(cb) {\r\n        callbacks.delete(cb);\r\n      };\r\n\r\n      event.__isCapturedEvent = true;\r\n\r\n      return triggerEvents;\r\n    }\r\n  }\n\n  function delay(ms) {\r\n    return new Promise((resolve) => {\r\n      setTimeout(resolve, ms);\r\n    });\r\n  }\n\n  /**\r\n   * Get matches from an object of nested objects\r\n   *\r\n   * @export\r\n   * @template T Type of matches\r\n   * @param {*} object Parent object to search\r\n   * @param {(x: any) => boolean} pred A predicate function that will receive each property value of an object\r\n   * @param {string[]} excludeKeys Exclude a property if the key exactly matches\r\n   * @returns {T[]} The matched values from the parent object\r\n   */\r\n  function getDeepMatches(object, pred, excludeKeys) {\r\n    const keys = typeof object === \"object\" ? Object.keys(object) : [];\r\n\r\n    return keys.length\r\n      ? keys\r\n          .filter((key) => !excludeKeys.includes(key))\r\n          .reduce((r, key) => {\r\n            const target = object[key];\r\n\r\n            if (target && pred(target)) {\r\n              return [...r, target];\r\n            } else {\r\n              return [...r, ...getDeepMatches(target, pred, excludeKeys)];\r\n            }\r\n          }, [] )\r\n      : [];\r\n  }\n\n  const importPath = /*@__PURE__*/JSON.parse(\"%PATH%\") ;\r\n  const delayLength = /*@__PURE__*/JSON.parse(\"%DELAY%\") ;\r\n  const excludedPaths = /*@__PURE__*/JSON.parse(\"%EXCLUDE%\") ;\n\n  const events = getDeepMatches(\r\n    chrome,\r\n    (x) => typeof x === \"object\" && \"addListener\" in x,\r\n    // The webRequest API is not compatible with event pages\r\n    //  TODO: this can be removed\r\n    //   if we stop using this wrapper with \"webRequest\" permission\r\n    excludedPaths.concat([\"webRequest\"]),\r\n  );\r\n  const triggerEvents = captureEvents(events);\r\n\r\n  import(importPath).then(async () => {\r\n    if (delayLength) await delay(delayLength);\r\n\r\n    triggerEvents();\r\n  });\n\n}());\n";

/**
 * This options object allows fine-tuning of the dynamic import wrapper.
 *
 * @export
 * @interface DynamicImportWrapper
 */









// FEATURE: add static code analysis for wake events
//  - This will be slower...
function prepImportWrapperScript({
  eventDelay = 0,
  wakeEvents = [],
  excludeNames = ["extension"],
}) {
  const delay = JSON.stringify(eventDelay);
  const events = wakeEvents.length
    ? JSON.stringify(
        wakeEvents.map((ev) => ev.replace(/^chrome\./, "")),
      )
    : false;
  const exclude = JSON.stringify(excludeNames);

  const script = (events
    ? code$4.replace("%EVENTS%", events)
    : code$3.replace("%EXCLUDE%", exclude)
  ).replace("%DELAY%", delay);

  return script;
}

const combinePerms = (
  ...permissions
) => {
  const { perms, xperms } = (permissions.flat(
    Infinity,
  ) )
    .filter((perm) => typeof perm !== "undefined")
    .reduce(
      ({ perms, xperms }, perm) => {
        if (perm.startsWith("!")) {
          xperms.add(perm.slice(1));
        } else {
          perms.add(perm);
        }

        return { perms, xperms };
      },
      { perms: new Set(), xperms: new Set() },
    );

  return [...perms].filter((p) => !xperms.has(p));
};

/* ============================================ */
/*               CHECK PERMISSIONS              */
/* ============================================ */

// export const debugger = s => /((chromep?)|(browser))[\s\n]*\.[\s\n]*debugger/.test(s)
// export const enterprise.deviceAttributes = s => /((chromep?)|(browser))[\s\n]*\.[\s\n]*enterprise\.deviceAttributes/.test(s)
// export const enterprise.hardwarePlatform = s => /((chromep?)|(browser))[\s\n]*\.[\s\n]*enterprise\.hardwarePlatform/.test(s)
// export const enterprise.platformKeys = s => /((chromep?)|(browser))[\s\n]*\.[\s\n]*enterprise\.platformKeys/.test(s)
// export const networking.config = s => /((chromep?)|(browser))[\s\n]*\.[\s\n]*networking\.config/.test(s)
// export const system.cpu = s => /((chromep?)|(browser))[\s\n]*\.[\s\n]*system\.cpu/.test(s)
// export const system.display = s => /((chromep?)|(browser))[\s\n]*\.[\s\n]*system\.display/.test(s)
// export const system.memory = s => /((chromep?)|(browser))[\s\n]*\.[\s\n]*system\.memory/.test(s)
// export const system.storage = s => /((chromep?)|(browser))[\s\n]*\.[\s\n]*system\.storage/.test(s)

const alarms = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*alarms/.test(s);

const bookmarks = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*bookmarks/.test(s);

const contentSettings = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*contentSettings/.test(s);

const contextMenus = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*contextMenus/.test(s);

const cookies = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*cookies/.test(s);

const declarativeContent = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*declarativeContent/.test(s);
const declarativeNetRequest = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*declarativeNetRequest/.test(s);
const declarativeWebRequest = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*declarativeWebRequest/.test(s);
const desktopCapture = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*desktopCapture/.test(s);
const displaySource = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*displaySource/.test(s);
const dns = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*dns/.test(s);
const documentScan = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*documentScan/.test(s);
const downloads = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*downloads/.test(s);
const experimental = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*experimental/.test(s);
const fileBrowserHandler = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*fileBrowserHandler/.test(s);
const fileSystemProvider = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*fileSystemProvider/.test(s);
const fontSettings = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*fontSettings/.test(s);
const gcm = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*gcm/.test(s);
const geolocation = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*geolocation/.test(s);
const history = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*history/.test(s);
const identity = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*identity/.test(s);
const idle = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*idle/.test(s);
const idltest = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*idltest/.test(s);
const management = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*management/.test(s);
const nativeMessaging = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*nativeMessaging/.test(s);
const notifications = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*notifications/.test(s);
const pageCapture = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*pageCapture/.test(s);
const platformKeys = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*platformKeys/.test(s);
const power = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*power/.test(s);
const printerProvider = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*printerProvider/.test(s);
const privacy = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*privacy/.test(s);
const processes = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*processes/.test(s);
const proxy = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*proxy/.test(s);
const sessions = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*sessions/.test(s);
const signedInDevices = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*signedInDevices/.test(s);
const storage = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*storage/.test(s);
const tabCapture = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*tabCapture/.test(s);
// export const tabs = s => /((chromep?)|(browser))[\s\n]*\.[\s\n]*tabs/.test(s)
const topSites = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*topSites/.test(s);
const tts = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*tts/.test(s);
const ttsEngine = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*ttsEngine/.test(s);
const unlimitedStorage = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*unlimitedStorage/.test(s);
const vpnProvider = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*vpnProvider/.test(s);
const wallpaper = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*wallpaper/.test(s);
const webNavigation = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*webNavigation/.test(s);
const webRequest = (s) =>
  /((chromep?)|(browser))[\s\n]*\.[\s\n]*webRequest/.test(s);
const webRequestBlocking = (s) =>
  webRequest(s) && s.includes("'blocking'");

// TODO: add readClipboard
// TODO: add writeClipboard

var permissions = /*#__PURE__*/Object.freeze({
  __proto__: null,
  alarms: alarms,
  bookmarks: bookmarks,
  contentSettings: contentSettings,
  contextMenus: contextMenus,
  cookies: cookies,
  declarativeContent: declarativeContent,
  declarativeNetRequest: declarativeNetRequest,
  declarativeWebRequest: declarativeWebRequest,
  desktopCapture: desktopCapture,
  displaySource: displaySource,
  dns: dns,
  documentScan: documentScan,
  downloads: downloads,
  experimental: experimental,
  fileBrowserHandler: fileBrowserHandler,
  fileSystemProvider: fileSystemProvider,
  fontSettings: fontSettings,
  gcm: gcm,
  geolocation: geolocation,
  history: history,
  identity: identity,
  idle: idle,
  idltest: idltest,
  management: management,
  nativeMessaging: nativeMessaging,
  notifications: notifications,
  pageCapture: pageCapture,
  platformKeys: platformKeys,
  power: power,
  printerProvider: printerProvider,
  privacy: privacy,
  processes: processes,
  proxy: proxy,
  sessions: sessions,
  signedInDevices: signedInDevices,
  storage: storage,
  tabCapture: tabCapture,
  topSites: topSites,
  tts: tts,
  ttsEngine: ttsEngine,
  unlimitedStorage: unlimitedStorage,
  vpnProvider: vpnProvider,
  wallpaper: wallpaper,
  webNavigation: webNavigation,
  webRequest: webRequest,
  webRequestBlocking: webRequestBlocking
});

/* ============================================ */
/*              DERIVE PERMISSIONS              */
/* ============================================ */

const derivePermissions = (
    set,
    { code },
) =>
    Object.entries(permissions)
        .filter(([, fn]) => fn(code))
        .map(([key]) => key)
        .reduce((s, p) => s.add(p), set);

// /* ============================================ */
// /*                DERIVE MANIFEST               */
// /* ============================================ */

// export function deriveManifest(
//   manifest: ChromeExtensionManifest, // manifest.json
//   ...permissions: string[] | string[][] // will be combined with manifest.permissions
// ): ChromeExtensionManifest {
//   return validateManifest({
//     // SMELL: Is this necessary?
//     manifest_version: 2,
//     ...manifest,
//     permissions: combinePerms(permissions, manifest.permissions),
//   })
// }

/* -------------------------------------------- */
/*                 DERIVE FILES                 */
/* -------------------------------------------- */

function deriveFiles(
    manifest,
    srcDir,
) {
    const files = get(
        manifest,
        "web_accessible_resources",
        [] ,
    ).reduce((resources, resourse) => resourse.resources.reduce((r, x) => {
        if (glob.hasMagic(x)) {
            const files = glob.sync(x, { cwd: srcDir });
            return [...r, ...files.map((f) => f.replace(srcDir, ""))];
        } else {
            return [...r, x];
        }
    }, resources), [] );

    const js = [
        ...files.filter((f) => /\.[jt]sx?$/.test(f)),
        get(manifest, "background.service_worker", "" ),
        ...get(
            manifest,
            "content_scripts",
            [] ,
        ).reduce((r, { js = [] }) => [...r, ...js], [] ),
    ];

    const html = [
        ...files.filter((f) => /\.html?$/.test(f)),
        get(manifest, "background.page"),
        get(manifest, "options_page"),
        get(manifest, "options_ui.page"),
        get(manifest, "devtools_page"),
        get(manifest, "browser_action.default_popup"),
        get(manifest, "page_action.default_popup"),
        ...Object.values(get(manifest, "chrome_url_overrides", {})),
    ];

    const css = [
        ...files.filter((f) => f.endsWith(".css")),
        ...get(
            manifest,
            "content_scripts",
            [] ,
        ).reduce(
            (r, { css = [] }) => [...r, ...css],
            [] ,
        ),
    ];

    // TODO: this can be a string or object
    const actionIconSet = [
        "browser_action.default_icon",
        "page_action.default_icon",
    ].reduce((set, query) => {
        const result = get(
            manifest,
            query,
            {},
        );

        if (typeof result === "string") {
            set.add(result);
        } else {
            Object.values(result).forEach((x) => set.add(x));
        }

        return set;
    }, new Set());

    const img = [
        ...actionIconSet,
        ...files.filter((f) =>
            /\.(jpe?g|png|svg|tiff?|gif|webp|bmp|ico)$/i.test(f),
        ),
        ...Object.values(get(manifest, "icons", {})),
    ];

    // Files like fonts, things that are not expected
    const others = diff(files, css, js, html, img);

    return {
        css: validate(css),
        js: validate(js),
        html: validate(html),
        img: validate(img),
        others: validate(others),
    };

    function validate(ary) {
        return [...new Set(ary.filter(isString))].map((x) =>
            join(srcDir, x),
        );
    }

    function isString(x) {
        return typeof x === "string";
    }
}

var id = "http://json-schema.org/draft-04/schema#";
var $schema$1 = "http://json-schema.org/draft-04/schema#";
var description = "Core schema meta-schema";
var definitions$1 = {
	schemaArray: {
		type: "array",
		minItems: 1,
		items: {
			$ref: "#"
		}
	},
	positiveInteger: {
		type: "integer",
		minimum: 0
	},
	positiveIntegerDefault0: {
		allOf: [
			{
				$ref: "#/definitions/positiveInteger"
			},
			{
				"default": 0
			}
		]
	},
	simpleTypes: {
		"enum": [
			"array",
			"boolean",
			"integer",
			"null",
			"number",
			"object",
			"string"
		]
	},
	stringArray: {
		type: "array",
		items: {
			type: "string"
		},
		minItems: 1,
		uniqueItems: true
	}
};
var type$1 = "object";
var properties$1 = {
	id: {
		type: "string"
	},
	$schema: {
		type: "string"
	},
	title: {
		type: "string"
	},
	description: {
		type: "string"
	},
	"default": {
	},
	multipleOf: {
		type: "number",
		minimum: 0,
		exclusiveMinimum: true
	},
	maximum: {
		type: "number"
	},
	exclusiveMaximum: {
		type: "boolean",
		"default": false
	},
	minimum: {
		type: "number"
	},
	exclusiveMinimum: {
		type: "boolean",
		"default": false
	},
	maxLength: {
		$ref: "#/definitions/positiveInteger"
	},
	minLength: {
		$ref: "#/definitions/positiveIntegerDefault0"
	},
	pattern: {
		type: "string",
		format: "regex"
	},
	additionalItems: {
		anyOf: [
			{
				type: "boolean"
			},
			{
				$ref: "#"
			}
		],
		"default": {
		}
	},
	items: {
		anyOf: [
			{
				$ref: "#"
			},
			{
				$ref: "#/definitions/schemaArray"
			}
		],
		"default": {
		}
	},
	maxItems: {
		$ref: "#/definitions/positiveInteger"
	},
	minItems: {
		$ref: "#/definitions/positiveIntegerDefault0"
	},
	uniqueItems: {
		type: "boolean",
		"default": false
	},
	maxProperties: {
		$ref: "#/definitions/positiveInteger"
	},
	minProperties: {
		$ref: "#/definitions/positiveIntegerDefault0"
	},
	required: {
		$ref: "#/definitions/stringArray"
	},
	additionalProperties: {
		anyOf: [
			{
				type: "boolean"
			},
			{
				$ref: "#"
			}
		],
		"default": {
		}
	},
	definitions: {
		type: "object",
		additionalProperties: {
			$ref: "#"
		},
		"default": {
		}
	},
	properties: {
		type: "object",
		additionalProperties: {
			$ref: "#"
		},
		"default": {
		}
	},
	patternProperties: {
		type: "object",
		additionalProperties: {
			$ref: "#"
		},
		"default": {
		}
	},
	dependencies: {
		type: "object",
		additionalProperties: {
			anyOf: [
				{
					$ref: "#"
				},
				{
					$ref: "#/definitions/stringArray"
				}
			]
		}
	},
	"enum": {
		type: "array",
		minItems: 1,
		uniqueItems: true
	},
	type: {
		anyOf: [
			{
				$ref: "#/definitions/simpleTypes"
			},
			{
				type: "array",
				items: {
					$ref: "#/definitions/simpleTypes"
				},
				minItems: 1,
				uniqueItems: true
			}
		]
	},
	format: {
		type: "string"
	},
	allOf: {
		$ref: "#/definitions/schemaArray"
	},
	anyOf: {
		$ref: "#/definitions/schemaArray"
	},
	oneOf: {
		$ref: "#/definitions/schemaArray"
	},
	not: {
		$ref: "#"
	}
};
var dependencies$1 = {
	exclusiveMaximum: [
		"maximum"
	],
	exclusiveMinimum: [
		"minimum"
	]
};
var jsonSchema = {
	id: id,
	$schema: $schema$1,
	description: description,
	definitions: definitions$1,
	type: type$1,
	properties: properties$1,
	dependencies: dependencies$1,
	"default": {
}
};

var title = "JSON schema for Google Chrome extension manifest files";
var $schema = "http://json-schema.org/draft-04/schema#";
var type = "object";
var additionalProperties = true;
var required = [
	"manifest_version",
	"name",
	"version"
];
var properties = {
	manifest_version: {
		type: "number",
		description: "One integer specifying the version of the manifest file format your package requires.",
		"enum": [
			3
		],
		minimum: 3,
		maximum: 3
	},
	name: {
		type: "string",
		description: "The name of the extension",
		maxLength: 45
	},
	version: {
		description: "One to four dot-separated integers identifying the version of this extension.",
		$ref: "#/definitions/version_string"
	},
	default_locale: {
		type: "string",
		description: "Specifies the subdirectory of _locales that contains the default strings for this extension.",
		"default": "en"
	},
	description: {
		type: "string",
		description: "A plain text description of the extension",
		maxLength: 132
	},
	icons: {
		type: "object",
		description: "One or more icons that represent the extension, app, or theme. Recommended format: PNG; also BMP, GIF, ICO, JPEG.",
		minProperties: 1,
		properties: {
			"16": {
				$ref: "#/definitions/icon",
				description: "Used as the favicon for an extension's pages and infobar."
			},
			"48": {
				$ref: "#/definitions/icon",
				description: "Used on the extension management page (chrome://extensions)."
			},
			"128": {
				$ref: "#/definitions/icon",
				description: "Used during installation and in the Chrome Web Store."
			},
			"256": {
				$ref: "#/definitions/icon",
				description: "Used during installation and in the Chrome Web Store."
			}
		}
	},
	browser_action: {
		$ref: "#/definitions/action",
		description: "Use browser actions to put icons in the main Google Chrome toolbar, to the right of the address bar. In addition to its icon, a browser action can also have a tooltip, a badge, and a popup."
	},
	page_action: {
		$ref: "#/definitions/action",
		description: "Use the chrome.pageAction API to put icons inside the address bar. Page actions represent actions that can be taken on the current page, but that aren't applicable to all pages."
	},
	background: {
		type: "object",
		description: "The background page is an HTML page that runs in the extension process. It exists for the lifetime of your extension, and only one instance of it at a time is active.",
		properties: {
			persistent: {
				type: "boolean",
				description: "When false, makes the background page an event page (loaded only when needed).",
				"default": true
			},
			page: {
				$ref: "#/definitions/page",
				description: "Specify the HTML of the background page.",
				"default": "background.html"
			},
			scripts: {
				$ref: "#/definitions/scripts",
				description: "A background page will be generated by the extension system that includes each of the files listed in the scripts property.",
				"default": [
					"background.js"
				]
			}
		},
		dependencies: {
			page: {
				not: {
					required: [
						"scripts"
					]
				}
			},
			scripts: {
				not: {
					required: [
						"page"
					]
				}
			}
		}
	},
	chrome_url_overrides: {
		type: "object",
		description: "Override pages are a way to substitute an HTML file from your extension for a page that Google Chrome normally provides.",
		additionalProperties: false,
		maxProperties: 1,
		properties: {
			bookmarks: {
				$ref: "#/definitions/page",
				description: "The page that appears when the user chooses the Bookmark Manager menu item from the Chrome menu or, on Mac, the Bookmark Manager item from the Bookmarks menu. You can also get to this page by entering the URL chrome://bookmarks.",
				"default": "bookmarks.html"
			},
			history: {
				$ref: "#/definitions/page",
				description: "The page that appears when the user chooses the History menu item from the Chrome menu or, on Mac, the Show Full History item from the History menu. You can also get to this page by entering the URL chrome://history.",
				"default": "history.html"
			},
			newtab: {
				$ref: "#/definitions/page",
				description: "The page that appears when the user creates a new tab or window. You can also get to this page by entering the URL chrome://newtab.",
				"default": "newtab.html"
			}
		}
	},
	commands: {
		type: "object",
		description: "Use the commands API to add keyboard shortcuts that trigger actions in your extension, for example, an action to open the browser action or send a command to the extension.",
		patternProperties: {
			".*": {
				$ref: "#/definitions/command"
			},
			"^_execute_browser_action$": {
				$ref: "#/definitions/command"
			},
			"^_execute_page_action$": {
				$ref: "#/definitions/command"
			}
		}
	},
	content_scripts: {
		type: "array",
		description: "Content scripts are JavaScript files that run in the context of web pages.",
		minItems: 1,
		uniqueItems: true,
		items: {
			type: "object",
			required: [
				"matches"
			],
			additionalProperties: false,
			properties: {
				matches: {
					type: "array",
					description: "Specifies which pages this content script will be injected into.",
					minItems: 1,
					uniqueItems: true,
					items: {
						$ref: "#/definitions/match_pattern"
					}
				},
				exclude_matches: {
					type: "array",
					description: "Excludes pages that this content script would otherwise be injected into.",
					uniqueItems: true,
					items: {
						$ref: "#/definitions/match_pattern"
					}
				},
				css: {
					type: "array",
					description: "The list of CSS files to be injected into matching pages. These are injected in the order they appear in this array, before any DOM is constructed or displayed for the page.",
					uniqueItems: true,
					items: {
						$ref: "#/definitions/uri"
					}
				},
				js: {
					$ref: "#/definitions/scripts",
					description: "The list of JavaScript files to be injected into matching pages. These are injected in the order they appear in this array."
				},
				run_at: {
					type: "string",
					description: "Controls when the files in js are injected.",
					"enum": [
						"document_start",
						"document_end",
						"document_idle"
					],
					"default": "document_idle"
				},
				all_frames: {
					type: "boolean",
					description: "Controls whether the content script runs in all frames of the matching page, or only the top frame.",
					"default": false
				},
				include_globs: {
					type: "array",
					description: "Applied after matches to include only those URLs that also match this glob. Intended to emulate the @include Greasemonkey keyword.",
					uniqueItems: true,
					items: {
						$ref: "#/definitions/glob_pattern"
					}
				},
				exclude_globs: {
					type: "array",
					description: "Applied after matches to exclude URLs that match this glob. Intended to emulate the @exclude Greasemonkey keyword.",
					uniqueItems: true,
					items: {
						$ref: "#/definitions/glob_pattern"
					}
				},
				match_about_blank: {
					type: "boolean",
					description: "Whether to insert the content script on about:blank and about:srcdoc.",
					"default": false
				}
			}
		}
	},
	content_security_policy: {
		$ref: "#/definitions/content_security_policy"
	},
	devtools_page: {
		$ref: "#/definitions/page",
		description: "A DevTools extension adds functionality to the Chrome DevTools. It can add new UI panels and sidebars, interact with the inspected page, get information about network requests, and more."
	},
	externally_connectable: {
		type: "object",
		description: "Declares which extensions, apps, and web pages can connect to your extension via runtime.connect and runtime.sendMessage.",
		items: {
			type: "object",
			additionalProperties: false,
			properties: {
				ids: {
					type: "array",
					items: {
						type: "string",
						description: "The IDs of extensions or apps that are allowed to connect. If left empty or unspecified, no extensions or apps can connect."
					}
				},
				matches: {
					type: "array",
					items: {
						type: "string",
						description: "The URL patterns for web pages that are allowed to connect. This does not affect content scripts. If left empty or unspecified, no web pages can connect."
					}
				},
				accepts_tls_channel_id: {
					type: "boolean",
					"default": false,
					description: "Indicates that the extension would like to make use of the TLS channel ID of the web page connecting to it. The web page must also opt to send the TLS channel ID to the extension via setting includeTlsChannelId to true in runtime.connect's connectInfo or runtime.sendMessage's options."
				}
			}
		}
	},
	file_browser_handlers: {
		type: "array",
		description: "You can use this API to enable users to upload files to your website.",
		minItems: 1,
		items: {
			type: "object",
			required: [
				"id",
				"default_title",
				"file_filters"
			],
			additionalProperties: false,
			properties: {
				id: {
					type: "string",
					description: "Used by event handling code to differentiate between multiple file handlers"
				},
				default_title: {
					type: "string",
					description: "What the button will display."
				},
				file_filters: {
					type: "array",
					description: "Filetypes to match.",
					minItems: 1,
					items: {
						type: "string"
					}
				}
			}
		}
	},
	homepage_url: {
		$ref: "#/definitions/uri",
		description: "The URL of the homepage for this extension."
	},
	incognito: {
		type: "string",
		description: "Specify how this extension will behave if allowed to run in incognito mode.",
		"enum": [
			"spanning",
			"split"
		],
		"default": "spanning"
	},
	input_components: {
		type: "array",
		description: "Allows your extension to handle keystrokes, set the composition, and manage the candidate window.",
		items: {
			type: "object",
			required: [
				"name",
				"type",
				"id",
				"description",
				"language",
				"layouts"
			],
			additionalProperties: false,
			properties: {
				name: {
					type: "string"
				},
				type: {
					type: "string"
				},
				id: {
					type: "string"
				},
				description: {
					type: "string"
				},
				language: {
					type: "string"
				},
				layouts: {
					type: "array"
				}
			}
		}
	},
	key: {
		type: "string",
		description: "This value can be used to control the unique ID of an extension, app, or theme when it is loaded during development."
	},
	minimum_chrome_version: {
		$ref: "#/definitions/version_string",
		description: "The version of Chrome that your extension, app, or theme requires, if any."
	},
	nacl_modules: {
		type: "array",
		description: "One or more mappings from MIME types to the Native Client module that handles each type.",
		minItems: 1,
		uniqueItems: true,
		items: {
			type: "object",
			required: [
				"path",
				"mime_type"
			],
			additionalProperties: false,
			properties: {
				path: {
					$ref: "#/definitions/uri",
					description: "The location of a Native Client manifest (a .nmf file) within the extension directory."
				},
				mime_type: {
					$ref: "#/definitions/mime_type",
					description: "The MIME type for which the Native Client module will be registered as content handler."
				}
			}
		}
	},
	oauth2: {
		type: "object",
		description: "Use the Chrome Identity API to authenticate users: the getAuthToken for users logged into their Google Account and the launchWebAuthFlow for users logged into a non-Google account.",
		required: [
			"client_id",
			"scopes"
		],
		additionalProperties: false,
		properties: {
			client_id: {
				type: "string",
				description: "You need to register your app in the Google APIs Console to get the client ID."
			},
			scopes: {
				type: "array",
				minItems: 1,
				items: {
					type: "string"
				}
			}
		}
	},
	offline_enabled: {
		type: "boolean",
		description: "Whether the app or extension is expected to work offline. When Chrome detects that it is offline, apps with this field set to true will be highlighted on the New Tab page."
	},
	omnibox: {
		type: "object",
		description: "The omnibox API allows you to register a keyword with Google Chrome's address bar, which is also known as the omnibox.",
		required: [
			"keyword"
		],
		additionalProperties: false,
		properties: {
			keyword: {
				type: "string",
				description: "The keyward that will trigger your extension."
			}
		}
	},
	optional_permissions: {
		$ref: "#/definitions/permissions",
		description: "Use the chrome.permissions API to request declared optional permissions at run time rather than install time, so users understand why the permissions are needed and grant only those that are necessary."
	},
	options_page: {
		$ref: "#/definitions/page",
		description: "To allow users to customize the behavior of your extension, you may wish to provide an options page. If you do, a link to it will be provided from the extensions management page at chrome://extensions. Clicking the Options link opens a new tab pointing at your options page.",
		"default": "options.html"
	},
	options_ui: {
		type: "object",
		description: "To allow users to customize the behavior of your extension, you may wish to provide an options page. If you do, an Options link will be shown on the extensions management page at chrome://extensions which opens a dialogue containing your options page.",
		required: [
			"page"
		],
		properties: {
			page: {
				type: "string",
				description: "The path to your options page, relative to your extension's root."
			},
			chrome_style: {
				type: "boolean",
				"default": true,
				description: "If true, a Chrome user agent stylesheet will be applied to your options page. The default value is false, but we recommend you enable it for a consistent UI with Chrome."
			},
			open_in_tab: {
				type: "boolean",
				"default": false,
				description: "If true, your extension's options page will be opened in a new tab rather than embedded in chrome://extensions. The default is false, and we recommend that you don't change it. This is only useful to delay the inevitable deprecation of the old options UI! It will be removed soon, so try not to use it. It will break."
			}
		}
	},
	permissions: {
		$ref: "#/definitions/permissions",
		description: "Permissions help to limit damage if your extension or app is compromised by malware. Some permissions are also displayed to users before installation, as detailed in Permission Warnings."
	},
	requirements: {
		type: "object",
		description: "Technologies required by the app or extension. Hosting sites such as the Chrome Web Store may use this list to dissuade users from installing apps or extensions that will not work on their computer.",
		additionalProperties: false,
		properties: {
			"3D": {
				type: "object",
				description: "The '3D' requirement denotes GPU hardware acceleration.",
				required: [
					"features"
				],
				additionalProperties: false,
				properties: {
					features: {
						type: "array",
						description: "List of the 3D-related features your app requires.",
						minItems: 1,
						uniqueItems: true,
						items: {
							type: "string",
							"enum": [
								"webgl"
							]
						}
					}
				}
			},
			plugins: {
				type: "object",
				description: "Indicates if an app or extension requires NPAPI to run. This requirement is enabled by default when the manifest includes the 'plugins' field.",
				required: [
					"npapi"
				],
				additionalProperties: false,
				properties: {
					npapi: {
						type: "boolean",
						"default": true
					}
				}
			}
		}
	},
	sandbox: {
		type: "object",
		description: "Defines an collection of app or extension pages that are to be served in a sandboxed unique origin, and optionally a Content Security Policy to use with them.",
		required: [
			"pages"
		],
		additionalProperties: false,
		properties: {
			pages: {
				type: "array",
				minItems: 1,
				uniqueItems: true,
				items: {
					$ref: "#/definitions/page"
				}
			},
			content_security_policy: {
				$ref: "#/definitions/content_security_policy",
				"default": "sandbox allow-scripts allow-forms"
			}
		}
	},
	short_name: {
		type: "string",
		description: "The short name is typically used where there is insufficient space to display the full name.",
		maxLength: 12
	},
	update_url: {
		$ref: "#/definitions/uri",
		description: "If you publish using the Chrome Developer Dashboard, ignore this field. If you bridge your own extension or app: URL to an update manifest XML file."
	},
	tts_engine: {
		type: "object",
		description: "Register itself as a speech engine.",
		required: [
			"voices"
		],
		additionalProperties: false,
		properties: {
			voices: {
				type: "array",
				description: "Voices the extension can synthesize.",
				minItems: 1,
				uniqueItems: true,
				items: {
					type: "object",
					required: [
						"voice_name",
						"event_types"
					],
					additionalProperties: false,
					properties: {
						voice_name: {
							type: "string",
							description: "Identifies the name of the voice and the engine used."
						},
						lang: {
							type: "string",
							description: "Almost always, a voice can synthesize speech in just a single language. When an engine supports more than one language, it can easily register a separate voice for each language."
						},
						gender: {
							type: "string",
							description: "If your voice corresponds to a male or female voice, you can use this parameter to help clients choose the most appropriate voice for their application."
						},
						event_types: {
							type: "array",
							description: "Events sent to update the client on the progress of speech synthesis.",
							minItems: 1,
							uniqueItems: true,
							items: {
								type: "string",
								description: "",
								"enum": [
									"start",
									"word",
									"sentence",
									"marker",
									"end",
									"error"
								]
							}
						}
					}
				}
			}
		}
	},
	version_name: {
		type: "string",
		description: "In addition to the version field, which is used for update purposes, version_name can be set to a descriptive version string and will be used for display purposes if present."
	},
	web_accessible_resources: {
		type: "array",
		description: "An array of strings specifying the paths (relative to the package root) of packaged resources that are expected to be usable in the context of a web page.",
		minItems: 1,
		uniqueItems: true,
		items: {
			type: "object",
			additionalProperties: false,
			properties: {
				resources: {
					type: "array",
					description: "",
					items: {
						type: "string"
					}
				},
				matches: {
					type: "array",
					description: "",
					items: {
						type: "string"
					}
				},
				extension_ids: {
					type: "array",
					description: "",
					items: {
						type: "string"
					}
				},
				use_dynamic_url: {
					type: "boolean",
					description: ""
				}
			}
		}
	},
	chrome_settings_overrides: {
	},
	content_pack: {
	},
	current_locale: {
	},
	"import": {
	},
	platforms: {
	},
	signature: {
	},
	spellcheck: {
	},
	storage: {
	},
	system_indicator: {
	}
};
var dependencies = {
	page_action: {
		not: {
			required: [
				"browser_action"
			]
		}
	},
	browser_action: {
		not: {
			required: [
				"page_action"
			]
		}
	},
	content_scripts: {
		not: {
			required: [
				"script_badge"
			]
		}
	},
	script_badge: {
		not: {
			required: [
				"content_scripts"
			]
		}
	}
};
var definitions = {
	action: {
		type: "object",
		properties: {
			default_title: {
				type: "string",
				description: "Tooltip for the main toolbar icon."
			},
			default_popup: {
				$ref: "#/definitions/uri",
				description: "The popup appears when the user clicks the icon."
			},
			default_icon: {
				anyOf: [
					{
						type: "string",
						description: "FIXME: String form is deprecated."
					},
					{
						type: "object",
						description: "Icon for the main toolbar.",
						properties: {
							"19": {
								$ref: "#/definitions/icon"
							},
							"38": {
								$ref: "#/definitions/icon"
							}
						}
					}
				]
			}
		},
		dependencies: {
			name: {
				not: {
					required: [
						"name"
					]
				}
			},
			icons: {
				not: {
					required: [
						"icons"
					]
				}
			},
			popup: {
				not: {
					required: [
						"popup"
					]
				}
			}
		}
	},
	command: {
		type: "object",
		additionalProperties: false,
		properties: {
			description: {
				type: "string"
			},
			suggested_key: {
				type: "object",
				additionalProperties: false,
				patternProperties: {
					"^(default|mac|windows|linux|chromeos)$": {
						type: "string",
						pattern: "^(Ctrl|Command|MacCtrl|Alt|Option)\\+(Shift\\+)?[A-Z]"
					}
				}
			}
		}
	},
	content_security_policy: {
		type: "string",
		description: "This introduces some fairly strict policies that will make extensions more secure by default, and provides you with the ability to create and enforce rules governing the types of content that can be loaded and executed by your extensions and applications.",
		"default": "script-src 'self'; object-src 'self'"
	},
	glob_pattern: {
		type: "string"
	},
	icon: {
		$ref: "#/definitions/uri"
	},
	match_pattern: {
		type: "string",
		pattern: "^((\\*|http|https|file|ftp|chrome-extension):\\/\\/(\\*|\\*\\.[^\\/\\*]+|[^\\/\\*]+)?(\\/.*))|<all_urls>$"
	},
	mime_type: {
		type: "string",
		pattern: "^(?:application|audio|image|message|model|multipart|text|video)\\/[-+.\\w]+$"
	},
	page: {
		$ref: "#/definitions/uri"
	},
	permissions: {
		type: "array",
		uniqueItems: true,
		items: {
			type: "string"
		}
	},
	scripts: {
		type: "array",
		minItems: 1,
		uniqueItems: true,
		items: {
			$ref: "#/definitions/uri"
		}
	},
	uri: {
		type: "string"
	},
	version_string: {
		type: "string",
		pattern: "^(?:\\d{1,5}\\.){0,3}\\d{1,5}$"
	}
};
var manifestSchema = {
	title: title,
	$schema: $schema,
	type: type,
	additionalProperties: additionalProperties,
	required: required,
	properties: properties,
	dependencies: dependencies,
	definitions: definitions
};

class ValidationError extends Error {
  constructor(msg, errors) {
    super(msg);
    this.name = "ValidationError";
    this.errors = errors;
  }
  
}

// const jsonSchema = readJSONSync(
//   resolve(__dirname, 'json-schema-draft-04.json'),
// )

// const manifestSchema = readJSONSync(
//   resolve(__dirname, 'schema-web-ext-manifest-v3.json'),
// )

const ajv = new Ajv({
  verbose: true,
  schemaId: "auto",
  schemas: {
    "http://json-schema.org/draft-04/schema#": jsonSchema,
  },
  strictDefaults: true,
});

// ajv.addMetaSchema(jsonSchema)

const validator = ajv.compile(manifestSchema);

const validateManifest = (
  manifest,
) => {
  if (validator(manifest)) {
    return manifest;
  }

  const { errors } = validator;
  const msg = "There were problems with the extension manifest.";

  throw new ValidationError(msg, errors);
};

function dedupe(x) {
    return [...new Set(x)];
}

const explorer = cosmiconfigSync("manifest", {
    cache: false,
});

const name = "manifest-input";

const stubChunkName = "stub__empty-chrome-extension-manifest";

const npmPkgDetails =
    process.env.npm_package_name &&
        process.env.npm_package_version &&
        process.env.npm_package_description
        ? {
            name: process.env.npm_package_name,
            version: process.env.npm_package_version,
            description: process.env.npm_package_description,
        }
        : {
            name: "",
            version: "",
            description: "",
        };

/* ============================================ */
/*                MANIFEST-INPUT                */
/* ============================================ */

function manifestInput(
    {
        browserPolyfill = false,
        contentScriptWrapper = true,
        crossBrowser = false,
        dynamicImportWrapper = {},
        extendManifest = {},
        firstClassManifest = true,
        iifeJsonPaths = [],
        pkg = npmPkgDetails,
        publicKey,
        verbose = true,
        cache = {
            assetChanged: false,
            assets: [],
            iife: [],
            input: [],
            inputAry: [],
            inputObj: {},
            permsHash: "",
            readFile: new Map(),
            srcDir: null,
        } ,
    } = {} ,
) {
    const readAssetAsBuffer = memoize(
        (filepath) => {
            return fs.readFile(filepath);
        },
        {
            cache: cache.readFile,
        },
    );

    /* ----------- HOOKS CLOSURES START ----------- */

    let manifestPath;

    const manifestName = "manifest.json";

    /* ------------ HOOKS CLOSURES END ------------ */

    /* - SETUP DYNAMIC IMPORT LOADER SCRIPT START - */

    let wrapperScript = "";
    if (dynamicImportWrapper !== false) {
        wrapperScript = prepImportWrapperScript(dynamicImportWrapper);
    }

    /* -- SETUP DYNAMIC IMPORT LOADER SCRIPT END -- */

    /* --------------- plugin object -------------- */
    return {
        name,

        browserPolyfill,
        crossBrowser,

        get srcDir() {
            return cache.srcDir;
        },

        get formatMap() {
            return { iife: cache.iife };
        },

        /* ============================================ */
        /*                 OPTIONS HOOK                 */
        /* ============================================ */

        options(options) {
            // Do not reload manifest without changes
            if (!cache.manifest) {
                /* ----------- LOAD AND PROCESS MANIFEST ----------- */

                let inputManifestPath;
                if (Array.isArray(options.input)) {
                    const manifestIndex = options.input.findIndex(
                        isJsonFilePath,
                    );
                    inputManifestPath = options.input[manifestIndex];
                    cache.inputAry = [
                        ...options.input.slice(0, manifestIndex),
                        ...options.input.slice(manifestIndex + 1),
                    ];
                } else if (typeof options.input === "object") {
                    inputManifestPath = options.input.manifest;
                    cache.inputObj = cloneObject(options.input);
                    delete cache.inputObj["manifest"];
                } else {
                    inputManifestPath = options.input;
                }

                if (!isJsonFilePath(inputManifestPath)) {
                    throw new TypeError(
                        "RollupOptions.input must be a single Chrome extension manifest.",
                    );
                }

                const configResult = explorer.load(
                    inputManifestPath,
                ) 



;

                if (configResult.isEmpty) {
                    throw new Error(`${options.input} is an empty file.`);
                }

                const { options_page, options_ui } = configResult.config;
                if (
                    options_page !== undefined &&
                    options_ui !== undefined
                ) {
                    throw new Error(
                        "options_ui and options_page cannot both be defined in manifest.json.",
                    );
                }

                manifestPath = configResult.filepath;

                if (typeof extendManifest === "function") {
                    cache.manifest = extendManifest(configResult.config);
                } else if (typeof extendManifest === "object") {
                    cache.manifest = {
                        ...configResult.config,
                        ...extendManifest,
                    };
                } else {
                    cache.manifest = configResult.config;
                }

                cache.srcDir = path.dirname(manifestPath);

                if (firstClassManifest) {
                    cache.iife = iifeJsonPaths
                        .map((jsonPath) => {
                            const result = JSONPath({
                                path: jsonPath,
                                json: cache.manifest,
                            });

                            return result;
                        })
                        .flat(Infinity);

                    // Derive entry paths from manifest
                    const { js, html, css, img, others } = deriveFiles(
                        cache.manifest,
                        cache.srcDir,
                    );

                    // Cache derived inputs
                    cache.input = [...cache.inputAry, ...js, ...html];

                    cache.assets = [
                        // Dedupe assets
                        ...new Set([...css, ...img, ...others]),
                    ];
                }

                /* --------------- END LOAD MANIFEST --------------- */
            }

            const finalInput = cache.input.reduce(
                reduceToRecord(cache.srcDir),
                cache.inputObj,
            );

            if (Object.keys(finalInput).length === 0) {
                finalInput[stubChunkName] = stubChunkName;
            }

            return { ...options, input: finalInput };
        },

        /* ============================================ */
        /*              HANDLE WATCH FILES              */
        /* ============================================ */

        async buildStart() {
            this.addWatchFile(manifestPath);

            cache.assets.forEach((srcPath) => {
                this.addWatchFile(srcPath);
            });

            const assets = await Promise.all(
                cache.assets.map(async (srcPath) => {
                    const source = await readAssetAsBuffer(srcPath);

                    return {
                        type: "asset" ,
                        source,
                        fileName: path.relative(cache.srcDir, srcPath),
                    };
                }),
            );

            assets.forEach((asset) => {
                this.emitFile(asset);
            });
        },

        resolveId(source) {
            if (source === stubChunkName) {
                return source;
            }

            return null;
        },

        load(id) {
            if (id === stubChunkName) {
                return { code: `console.log(${stubChunkName})` };
            }

            return null;
        },

        watchChange(id) {
            if (id.endsWith(manifestName)) {
                // Dump cache.manifest if manifest changes
                delete cache.manifest;
                cache.assetChanged = false;
            } else {
                // Force new read of changed asset
                cache.assetChanged = cache.readFile.delete(id);
            }
        },

        /* ============================================ */
        /*                GENERATEBUNDLE                */
        /* ============================================ */

        generateBundle(options, bundle) {
            /* ----------------- CLEAN UP STUB ----------------- */

            delete bundle[stubChunkName + ".js"];

            /* ---------- DERIVE PERMISSIONS START --------- */

            // Get module ids for all chunks
            let permissions;
            if (cache.assetChanged && cache.permsHash) {
                // Permissions did not change
                permissions = JSON.parse(cache.permsHash) ;

                cache.assetChanged = false;
            } else {
                const chunks = Object.values(bundle).filter(isChunk);

                // Permissions may have changed
                permissions = Array.from(
                    chunks.reduce(derivePermissions, new Set()),
                );

                const permsHash = JSON.stringify(permissions);

                if (verbose && permissions.length) {
                    if (!cache.permsHash) {
                        this.warn(
                            `Detected permissions: ${permissions.toString()}`,
                        );
                    } else if (permsHash !== cache.permsHash) {
                        this.warn(
                            `Detected new permissions: ${permissions.toString()}`,
                        );
                    }
                }

                cache.permsHash = permsHash;
            }

            if (Object.keys(bundle).length === 0) {
                throw new Error(
                    "The manifest must have at least one asset (html or css) or script file.",
                );
            }

            try {
                // Clone cache.manifest
                if (!cache.manifest)
                    // This is a programming error, so it should throw
                    throw new TypeError(
                        `cache.manifest is ${typeof cache.manifest}`,
                    );

                const clonedManifest = cloneObject(cache.manifest);

                const manifestBody = validateManifest({
                    manifest_version: 3,
                    name: pkg.name,
                    version: pkg.version,
                    description: pkg.description,
                    ...clonedManifest,
                    permissions: combinePerms(
                        permissions,
                        clonedManifest.permissions || [],
                    ),
                });

                const {
                    content_scripts: cts = [],
                    web_accessible_resources: war = [],
                    background: { scripts: bgs = [] } = {},
                } = manifestBody;

                /* ------------- SETUP CONTENT SCRIPTS ------------- */

                const contentScripts = cts.reduce(
                    (r, { js = [] }) => [...r, ...js],
                    [] ,
                );

                if (contentScriptWrapper && contentScripts.length) {
                    const memoizedEmitter = memoize(
                        (scriptPath) => {
                            const source = code$5.replace(
                                "%PATH%",
                                // Fix path slashes to support Windows
                                JSON.stringify(
                                    slash(relative("assets", scriptPath)),
                                ),
                            );

                            const assetId = this.emitFile({
                                type: "asset",
                                source,
                                name: basename(scriptPath),
                            });

                            return this.getFileName(assetId);
                        },
                    );

                    // Setup content script import wrapper
                    manifestBody.content_scripts = cts.map(
                        ({ js, ...rest }) => {
                            return typeof js === "undefined"
                                ? rest
                                : {
                                    js: js
                                        .map(normalizeFilename)
                                        .map(memoizedEmitter)
                                        .map((p) => slash(p)),
                                    ...rest,
                                };
                        },
                    );

                    // make all imports & dynamic imports web_acc_res
                    const imports = Object.values(bundle)
                        .filter((x) => x.type === "chunk")
                        .reduce(
                            (r, { isEntry, fileName }) =>
                                // Get imported filenames
                                !isEntry ? [...r, fileName] : r,
                            [] ,
                        );

                    // SMELL: web accessible resources can be used for fingerprinting extensions
                    // TODO 这里逻辑被屏蔽 需要更新
                    const webAccessResourcesFileNames = dedupe([
                        // FEATURE: filter out imports for background?
                        ...imports,
                        // Need to be web accessible b/c of import
                        ...contentScripts,
                    ]).map((p) => slash(p));
                    // origin web_access_resources
                    console.log(war);
                    console.log(webAccessResourcesFileNames);
                    manifestBody.web_accessible_resources = [];
                }

                /* ----------- END SETUP CONTENT SCRIPTS ----------- */

                /* ------------ SETUP BACKGROUND SCRIPTS ----------- */

                // Emit background script wrappers
                if (bgs.length && wrapperScript.length) {
                    // background exists because bgs has scripts
                    manifestBody.background.scripts = bgs
                        .map(normalizeFilename)
                        .map((scriptPath) => {
                            // Loader script exists because of type guard above
                            const source =
                                // Path to module being loaded
                                wrapperScript.replace(
                                    "%PATH%",
                                    // Fix path slashes to support Windows
                                    JSON.stringify(
                                        slash(relative("assets", scriptPath)),
                                    ),
                                );

                            const assetId = this.emitFile({
                                type: "asset",
                                source,
                                name: basename(scriptPath),
                            });

                            return this.getFileName(assetId);
                        })
                        .map((p) => slash(p));
                }

                /* ---------- END SETUP BACKGROUND SCRIPTS --------- */

                /* --------- STABLE EXTENSION ID BEGIN -------- */

                if (publicKey) {
                    manifestBody.key = publicKey;
                }

                /* ---------- STABLE EXTENSION ID END --------- */

                /* ----------- OUTPUT MANIFEST.JSON BEGIN ---------- */

                const manifestJson = JSON.stringify(
                    manifestBody,
                    null,
                    2,
                )
                    // SMELL: is this necessary?
                    .replace(/\.[jt]sx?"/g, '.js"');

                // Emit manifest.json
                this.emitFile({
                    type: "asset",
                    fileName: manifestName,
                    source: manifestJson,
                });
            } catch (error) {
                // Catch here because we need the validated result in scope

                if (error.name !== "ValidationError") throw error;
                const errors = error.errors ;
                if (errors) {
                    errors.forEach((err) => {
                        // FIXME: make a better validation error message
                        // https://github.com/atlassian/better-ajv-errors
                        this.warn(JSON.stringify(err, undefined, 2));
                    });
                }
                this.error(error.message);
            }

            /* ------------ OUTPUT MANIFEST.JSON END ----------- */
        },
    };
}

const code$2 = "(function () {\n  'use strict';\n\n  /* eslint-disable semi */\r\n  const checkPolyfilled = 'typeof browser !== \"undefined\"';\r\n\r\n  const _executeScript = chrome.tabs.executeScript;\r\n  const withP = (...args) =>\r\n    new Promise((resolve, reject) => {\r\n      // @ts-ignore\r\n      _executeScript(...args, (results) => {\r\n        if (chrome.runtime.lastError) {\r\n          reject(chrome.runtime.lastError.message);\r\n        } else {\r\n          resolve(results);\r\n        }\r\n      });\r\n    });\r\n\r\n  chrome.tabs.executeScript = (...args) => {\r\n  (async () => {\r\n      const baseArgs = (typeof args[0] === 'number' ? [args[0]] : []); \r\n\r\n      const [done] = await withP(...(baseArgs.concat({ code: checkPolyfilled }) ));\r\n\r\n      if (!done) {\r\n        await withP(...(baseArgs.concat([{ file: JSON.parse('%BROWSER_POLYFILL_PATH%') }]) ));\r\n      }\r\n\r\n      _executeScript(...(args ));\r\n    })();\r\n  };\n\n}());\n";

function _optionalChain$1(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }





const defaultOptions = { executeScript: true };
function browserPolyfill({
  browserPolyfill: options = defaultOptions,
})


 {
  if (options === false)
    return {
      name: "no-op",
      generateBundle() {},
    };
  else if (options === true) options = defaultOptions;
  const { executeScript = true } = options;

  const convert = require("convert-source-map");
  const polyfillPath = require.resolve("webextension-polyfill");
  const src = fs.readFileSync(polyfillPath, "utf-8");
  const map = fs.readJsonSync(polyfillPath + ".map");

  const browserPolyfillSrc = [
    convert.removeMapFileComments(src),
    convert.fromObject(map).toComment(),
  ].join("\n");

  return {
    name: "browser-polyfill",
    generateBundle({ plugins = [] }, bundle) {
      const firefoxPlugin = plugins.find(
        ({ name }) => name === "firefox-addon",
      );
      const chromeExtensionPlugin = plugins.find(
        ({ name }) => name === "chrome-extension",
      ) ;

      if (
        firefoxPlugin &&
        !chromeExtensionPlugin._plugins.manifest.crossBrowser
      ) {
        return; // Don't need to add it
      }

      const manifestAsset = bundle["manifest.json"];
      if (!isAsset(manifestAsset)) {
        throw new TypeError(
          `manifest.json must be an OutputAsset, received "${typeof manifestAsset}"`,
        );
      }
      const manifest = JSON.parse(
        manifestAsset.source ,
      ) ;

      /* ------------- EMIT BROWSER POLYFILL ------------- */

      const bpId = this.emitFile({
        type: "asset",
        source: browserPolyfillSrc,
        fileName: "assets/browser-polyfill.js",
      });

      const browserPolyfillPath = this.getFileName(bpId);

      if (executeScript) {
        const exId = this.emitFile({
          type: "asset",
          source: code$2.replace(
            "%BROWSER_POLYFILL_PATH%",
            JSON.stringify(browserPolyfillPath),
          ),
          fileName: "assets/browser-polyfill-executeScript.js",
        });

        const executeScriptPolyfillPath = this.getFileName(exId);

        _optionalChain$1([manifest, 'access', _ => _.background, 'optionalAccess', _2 => _2.scripts, 'optionalAccess', _3 => _3.unshift, 'call', _4 => _4(
          executeScriptPolyfillPath,
        )]);
      }

      _optionalChain$1([manifest, 'access', _5 => _5.background, 'optionalAccess', _6 => _6.scripts, 'optionalAccess', _7 => _7.unshift, 'call', _8 => _8(browserPolyfillPath)]);
      _optionalChain$1([manifest, 'access', _9 => _9.content_scripts, 'optionalAccess', _10 => _10.forEach, 'call', _11 => _11((script) => {
        _optionalChain$1([script, 'access', _12 => _12.js, 'optionalAccess', _13 => _13.unshift, 'call', _14 => _14(browserPolyfillPath)]);
      })]);

      /* ---------------- UPDATE MANIFEST ---------------- */
      manifestAsset.source = JSON.stringify(manifest);
    },
  };
}

const validateNames = () => ({
  name: "validate-names",

  generateBundle(options, bundle) {
    const chunks = Object.values(bundle).filter(
      (x) => x.type === "chunk",
    );

    // Files cannot start with "_" in Chrome Extensions
    // Loop through each file and check for "_" in filename
    Object.keys(bundle)
      .filter((fileName) => fileName.startsWith("_"))
      .forEach((fileName) => {
        // Only replace first instance
        const regex = new RegExp(fileName);
        const fixed = fileName.slice(1);

        // Fix manifest
        const manifest = bundle["manifest.json"] ;
        manifest.source = manifest.source.replace(regex, fixed);

        // Change bundle key
        const chunk = bundle[fileName];
        delete bundle[fileName];
        bundle[fixed] = chunk;

        // Fix chunk
        chunk.fileName = fixed;

        // Find imports and fix
        chunks
          .filter(({ imports }) => imports.includes(fileName))
          .forEach((chunk) => {
            // Fix imports list
            chunk.imports = chunk.imports.map((i) =>
              i === fileName ? fixed : i,
            );
            // Fix imports in code
            chunk.code = chunk.code.replace(regex, fixed);
          });
      });
  },
});

const resolveFromBundle = (
  bundle,
) => ({
  name: "resolve-from-bundle",
  resolveId(source, importer) {
    if (typeof importer === "undefined") {
      return source;
    } else {
      const dirname = path.dirname(importer);
      const resolved = path.join(dirname, source);

      // if it's not in the bundle,
      //   tell Rollup not to try to resolve it
      return resolved in bundle ? resolved : false;
    }
  },
  load(id) {
    const chunk = bundle[id];

    if (isChunk(chunk)) {
      return {
        code: chunk.code,
        map: chunk.map,
      };
    } else {
      // anything not in the bundle is external
      //  this doesn't make sense for a chrome extension,
      //    but we should let Rollup handle it
      return null;
    }
  },
});

async function regenerateBundle(
  
  { input, output },
  bundle,
) {
  if (!output || Array.isArray(output)) {
    throw new TypeError(
      "options.output must be an OutputOptions object",
    );
  }

  if (typeof input === "undefined") {
    throw new TypeError(
      "options.input should be an object, string array or string",
    );
  }

  // Don't do anything if input is an empty array
  if (Array.isArray(input) && input.length === 0) {
    return {};
  }

  const { format, chunkFileNames: cfn = "", sourcemap } = output;

  const cfnString = typeof cfn === "string" ? cfn : "main";

  const chunkFileNames = path.join(path.dirname(cfnString), "[name].js");

  // Transform input array to input object
  const inputValue = Array.isArray(input)
    ? input.reduce((r, x) => {
        const { dir, name } = path.parse(x);
        return { ...r, [path.join(dir, name)]: x };
      }, {} )
    : input;

  const build = await rollup({
    input: inputValue,
    plugins: [resolveFromBundle(bundle)],
  });

  let _b;
  await build.generate({
    format,
    sourcemap,
    chunkFileNames,
    plugins: [
      {
        name: "get-bundle",
        generateBundle(o, b) {
          _b = b;
        },
      } ,
    ],
  });
  const newBundle = _b;

  if (typeof inputValue === "string") {
    delete bundle[inputValue];

    const bundleKey = path.basename(inputValue);

    return {
      [inputValue]: {
        ...(newBundle[bundleKey] ),
        fileName: inputValue,
      },
    };
  } else {
    // Remove regenerated entries from bundle
    Object.values(inputValue).forEach((key) => {
      delete bundle[key];
    });

    return newBundle;
  }
}

function mixedFormat(
  options,
) {
  return {
    name: 'mixed-format',
    async generateBundle(
      
      { format, chunkFileNames, sourcemap },
      bundle,
    ) {
      const { formatMap } = options; // this might not be defined upon init

      if (typeof formatMap === 'undefined') return

      const formats = Object.entries(formatMap).filter(
        (
          x,
        


) => typeof x[1] !== 'undefined',
      );

      {
        const allInput = formats.flatMap(([, inputs]) =>
          Array.isArray(inputs)
            ? inputs
            : Object.values(inputs || {}),
        );
        const allInputSet = new Set(allInput);
        if (allInput.length !== allInputSet.size) {
          throw new Error(
            'formats should not have duplicate inputs',
          )
        }
      }

      // TODO: handle different kinds of formats differently?
      const bundles = await Promise.all(
        // Configured formats
        formats.flatMap(([f, inputs]) =>
          (Array.isArray(inputs)
            ? inputs
            : Object.values(inputs)
          ).map((input) =>
            regenerateBundle.call(
              this,
              {
                input,
                output: {
                  format: f,
                  chunkFileNames,
                  sourcemap,
                },
              },
              bundle,
            ),
          ),
        ),
      );

      // Base format (ESM)
      const base = await regenerateBundle.call(
        this,
        {
          input: Object.entries(bundle)
            .filter(([, file]) => isChunk(file) && file.isEntry)
            .map(([key]) => key),
          output: { format, chunkFileNames, sourcemap },
        },
        bundle,
      );

      // Empty bundle
      Object.entries(bundle)
        .filter(([, v]) => isChunk(v))
        .forEach(([key]) => {
          delete bundle[key];
        });

      // Refill bundle
      Object.assign(bundle, base, ...bundles);
    },
  }
}

const code$1 = "(function () {\n  'use strict';\n\n  /* ------------------- FILENAMES ------------------- */\r\n\r\n  /* ------------------ PLACEHOLDERS ----------------- */\r\n\r\n  const timestampPathPlaceholder = \"%TIMESTAMP_PATH%\";\r\n  const loadMessagePlaceholder = \"%LOAD_MESSAGE%\";\r\n  const ctScriptPathPlaceholder = \"%CONTENT_SCRIPT_PATH%\";\r\n  const unregisterServiceWorkersPlaceholder = \"%UNREGISTER_SERVICE_WORKERS%\";\r\n  const executeScriptPlaceholder = \"%EXECUTE_SCRIPT%\";\n\n  /* eslint-env browser */\r\n\r\n  // Log load message to browser dev console\r\n  console.log(loadMessagePlaceholder.slice(1, -1));\r\n\r\n  const options = {\r\n    executeScript: JSON.parse(executeScriptPlaceholder),\r\n    unregisterServiceWorkers: JSON.parse(\r\n      unregisterServiceWorkersPlaceholder,\r\n    ),\r\n  };\r\n\r\n  /* ---------- POLYFILL TABS.EXECUTESCRIPT ---------- */\r\n\r\n  if (options.executeScript) {\r\n    const markerId =\r\n      'rollup-plugin-chrome-extension-simple-reloader';\r\n\r\n    const addMarker = `{\r\n    const tag = document.createElement('meta');\r\n    tag.id = '${markerId}';\r\n    document.head.append(tag);\r\n  }`;\r\n\r\n    const checkMarker = `\r\n  !!document.head.querySelector('#${markerId}')\r\n  `;\r\n\r\n    // Modify chrome.tabs.executeScript to inject reloader\r\n    const _executeScript = chrome.tabs.executeScript;\r\n    const withP = (...args) =>\r\n      new Promise((resolve, reject) => {\r\n        // eslint-disable-next-line\r\n        // @ts-ignore\r\n        _executeScript(...args, (results) => {\r\n          if (chrome.runtime.lastError) {\r\n            reject(chrome.runtime.lastError.message);\r\n          } else {\r\n            resolve(results);\r\n          }\r\n        });\r\n      });\r\n\r\n    chrome.tabs.executeScript = (...args) => {\r\n  (async () => {\r\n        const tabId = typeof args[0] === 'number' ? args[0] : null;\r\n        const argsBase = (tabId === null ? [] : [tabId]); \r\n\r\n        const [done] = await withP(\r\n          ...(argsBase.concat({ code: checkMarker }) \r\n\r\n\r\n  ),\r\n        );\r\n\r\n        // Don't add reloader if it's already there\r\n        if (!done) {\r\n          await withP(\r\n            ...(argsBase.concat({ code: addMarker }) \r\n\r\n\r\n  ),\r\n          );\r\n\r\n          // execute reloader\r\n          const reloaderArgs = argsBase.concat([\r\n            // TODO: convert to file to get replacements right\r\n            { file: JSON.parse(ctScriptPathPlaceholder) },\r\n          ]); \r\n\r\n          await withP(...reloaderArgs);\r\n        }\r\n\r\n        _executeScript(...(args ));\r\n      })();\r\n    };\r\n  }\r\n\r\n  /* ----------- UNREGISTER SERVICE WORKERS ---------- */\r\n\r\n  if (options.unregisterServiceWorkers) {\r\n    // Modify chrome.runtime.reload to unregister sw's\r\n    const _runtimeReload = chrome.runtime.reload;\r\n    chrome.runtime.reload = () => {\r\n  (async () => {\r\n        await unregisterServiceWorkers();\r\n        _runtimeReload();\r\n      })();\r\n    };\r\n  }\r\n\r\n  async function unregisterServiceWorkers() {\r\n    try {\r\n      const registrations = await navigator.serviceWorker.getRegistrations();\r\n      await Promise.all(registrations.map((r) => r.unregister()));\r\n    } catch (error) {\r\n      console.error(error);\r\n    }\r\n  }\r\n\r\n  /* -------------- CHECK TIMESTAMP.JSON ------------- */\r\n\r\n  let timestamp;\r\n\r\n  const id = setInterval(async () => {\r\n    const t = await fetch(timestampPathPlaceholder)\r\n      .then((res) => {\r\n        localStorage.removeItem('chromeExtensionReloaderErrors');\r\n        return res.json()\r\n      })\r\n      .catch(handleFetchError);\r\n\r\n    if (typeof timestamp === 'undefined') {\r\n      timestamp = t;\r\n    } else if (timestamp !== t) {\r\n      chrome.runtime.reload();\r\n    }\r\n\r\n    function handleFetchError(error) {\r\n      clearInterval(id);\r\n\r\n      const errors =\r\n        localStorage.chromeExtensionReloaderErrors || 0;\r\n\r\n      if (errors < 5) {\r\n        localStorage.chromeExtensionReloaderErrors = errors + 1;\r\n\r\n        // Should reload at least once if fetch fails.\r\n        // The fetch will fail if the timestamp file is absent,\r\n        // thus the new build does not include the reloader\r\n        return 0\r\n      } else {\r\n        console.log(\r\n          'rollup-plugin-chrome-extension simple reloader error:',\r\n        );\r\n        console.error(error);\r\n\r\n        return timestamp\r\n      }\r\n    }\r\n  }, 1000);\n\n}());\n";

const code = "(function () {\n  'use strict';\n\n  /* ------------------- FILENAMES ------------------- */\r\n  const loadMessagePlaceholder = \"%LOAD_MESSAGE%\";\n\n  /* eslint-env browser */\r\n\r\n  // Log load message to browser dev console\r\n  console.log(loadMessagePlaceholder.slice(1, -1));\r\n\r\n  const { name } = chrome.runtime.getManifest();\r\n\r\n  const reload = () => {\r\n    console.log(`${name} has reloaded...`);\r\n\r\n    setTimeout(() => {\r\n      location.reload();\r\n    }, 500);\r\n  };\r\n\r\n  setInterval(() => {\r\n    try {\r\n      chrome.runtime.getManifest();\r\n    } catch (error) {\r\n      if (error.message === \"Extension context invalidated.\") {\r\n        reload();\r\n      }\r\n    }\r\n  }, 1000);\n\n}());\n";

/* ------------------- FILENAMES ------------------- */

const backgroundPageReloader =
  "background-page-reloader.js";
const contentScriptReloader = "content-script-reloader.js";
const timestampFilename = "timestamp.json";

/* ------------------ PLACEHOLDERS ----------------- */

const timestampPathPlaceholder = "%TIMESTAMP_PATH%";
const loadMessagePlaceholder = "%LOAD_MESSAGE%";
const ctScriptPathPlaceholder = "%CONTENT_SCRIPT_PATH%";
const unregisterServiceWorkersPlaceholder = "%UNREGISTER_SERVICE_WORKERS%";
const executeScriptPlaceholder = "%EXECUTE_SCRIPT%";

function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }


















// Used for testing
const _internalCache = {};

const simpleReloader = (
  {
    executeScript = true,
    unregisterServiceWorkers = true,
  } = {} ,
  cache = {} ,
) => {
  if (!process.env.ROLLUP_WATCH) {
    return undefined;
  }

  return {
    name: "chrome-extension-simple-reloader",

    generateBundle({ dir }, bundle) {
      const date = new Date();
      const time = `${date
        .getFullYear()
        .toString()
        .padStart(2, "0")}-${(date.getMonth() + 1)
        .toString()
        .padStart(2, "0")}-${date
        .getDate()
        .toString()
        .padStart(2, "0")} ${date
        .getHours()
        .toString()
        .padStart(2, "0")}:${date
        .getMinutes()
        .toString()
        .padStart(2, "0")}:${date
        .getSeconds()
        .toString()
        .padStart(2, "0")}`;

      cache.outputDir = dir;
      cache.loadMessage = [
        "DEVELOPMENT build with simple auto-reloader",
        `[${time}] waiting for changes...`,
      ].join("\n");

      /* --------------- EMIT CLIENT FILES --------------- */

      const emit = (
        name,
        source,
        isFileName,
      ) => {
        const id = this.emitFile({
          type: "asset",
          [isFileName ? "fileName" : "name"]: name,
          source,
        });

        return this.getFileName(id);
      };

      cache.timestampPath = emit(
        timestampFilename,
        JSON.stringify(Date.now()),
        true,
      );

      cache.ctScriptPath = emit(
        contentScriptReloader,
        code.replace(
          loadMessagePlaceholder,
          JSON.stringify(cache.loadMessage),
        ),
      );

      cache.bgScriptPath = emit(
        backgroundPageReloader,
        code$1
          .replace(timestampPathPlaceholder, cache.timestampPath)
          .replace(
            loadMessagePlaceholder,
            JSON.stringify(cache.loadMessage),
          )
          .replace(
            ctScriptPathPlaceholder,
            JSON.stringify(cache.ctScriptPath),
          )
          .replace(
            executeScriptPlaceholder,
            JSON.stringify(executeScript),
          )
          .replace(
            unregisterServiceWorkersPlaceholder,
            JSON.stringify(unregisterServiceWorkers),
          ),
      );

      // Update the exported cache
      Object.assign(_internalCache, cache);

      /* ---------------- UPDATE MANIFEST ---------------- */

      updateManifest(
        (manifest) => {
          /* ------------------ DESCRIPTION ------------------ */

          manifest.description = cache.loadMessage;

          /* ---------------- BACKGROUND PAGE ---------------- */

          if (!manifest.background) {
            manifest.background = {};
          }

          manifest.background.persistent = true;

          const { scripts: bgScripts = [] } = manifest.background;

          if (cache.bgScriptPath) {
            manifest.background.scripts = [
              cache.bgScriptPath,
              ...bgScripts,
            ];
          } else {
            this.error(
              `cache.bgScriptPath is ${typeof cache.bgScriptPath}`,
            );
          }

          /* ---------------- CONTENT SCRIPTS ---------------- */

          const { content_scripts: ctScripts } = manifest;

          if (cache.ctScriptPath) {
            manifest.content_scripts = _optionalChain([ctScripts, 'optionalAccess', _ => _.map, 'call', _2 => _2(
              ({ js = [], ...rest }) => ({
                js: [cache.ctScriptPath, ...js],
                ...rest,
              }),
            )]);
          } else {
            this.error(
              `cache.ctScriptPath is ${typeof cache.ctScriptPath}`,
            );
          }

          return manifest;
        },
        bundle,
        this.error,
      );

      // We'll write this file ourselves, we just need a safe path to write the timestamp
      delete bundle[cache.timestampPath];
    },

    /* -------------- WRITE TIMESTAMP FILE ------------- */
    async writeBundle() {
      try {
        await outputJson(
          join(cache.outputDir, cache.timestampPath),
          Date.now(),
        );
      } catch (err) {
        if (typeof err.message === "string") {
          this.error(
            `Unable to update timestamp file:\n\t${err.message}`,
          );
        } else {
          this.error("Unable to update timestamp file");
        }
      }
    },
  };
};

const chromeExtension = (
    options = {} ,
) => {
    /* --------------- LOAD PACKAGE.JSON --------------- */

    try {
        const packageJsonPath = join(process.cwd(), "package.json");
        options.pkg = options.pkg || readJSONSync(packageJsonPath);
    } catch (error) { }

    /* ----------------- SETUP PLUGINS ----------------- */

    const manifest = manifestInput(options);
    const html = htmlInputs(manifest);
    const validate = validateNames();
    const browser = browserPolyfill(manifest);
    const mixedFormat$1 = mixedFormat(manifest);

    /* ----------------- RETURN PLUGIN ----------------- */

    return {
        name: "chrome-extension",

        // For testing
        _plugins: { manifest, html, validate },

        async options(options) {
            try {
                return await [manifest, html].reduce(async (opts, plugin) => {
                    // eslint-disable-next-line @typescript-eslint/await-thenable
                    const result = await plugin.options.call(this, await opts);
                    return result || options;
                }, Promise.resolve(options));
            } catch (error) {
                const manifestError =
                    "The manifest must have at least one script or HTML file.";
                const htmlError =
                    "At least one HTML file must have at least one script.";

                if (
                    error.message === manifestError ||
                    error.message === htmlError
                ) {
                    throw new Error(
                        "A Chrome extension must have at least one script or HTML file.",
                    );
                } else {
                    throw error;
                }
            }
        },

        async buildStart(options) {
            await Promise.all([
                manifest.buildStart.call(this, options),
                html.buildStart.call(this, options),
            ]);
        },

        async resolveId(source, importer) {
            return manifest.resolveId.call(this, source, importer, {});
        },

        async load(id) {
            return manifest.load.call(this, id);
        },

        watchChange(id) {
            manifest.watchChange.call(this, id, { event: "create" });
            html.watchChange.call(this, id, { event: "create" });
        },

        async generateBundle(...args) {
            await manifest.generateBundle.call(this, ...args);
            await validate.generateBundle.call(this, ...args);
            await browser.generateBundle.call(this, ...args);
            // TODO: should skip this if not needed
            await mixedFormat$1.generateBundle.call(this, ...args);
        },
    };
};

export { chromeExtension, simpleReloader };
